File: Data Structures & Algorithms/products-of-array-discluding-self/submission-1.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        ret = [0] * len(nums)

        # first pass
        i = 0
        while i < len(nums):
            if i == 0:
                ret[i] = 1
            else:
                ret[i] = nums[i - 1] * ret[i - 1]
            
            i += 1
        
        # second pass:
        i = len(nums) - 1
        post = 1
        while i >= 0:
            if i == len(nums) - 1:
                ret[i] *= 1
            else:
                post *= nums[i + 1]
                ret[i] *= post

            i -= 1

        return ret
        

File: Data Structures & Algorithms/largest-rectangle-in-histogram/submission-1.cpp
class Solution {
public:
    int largestRectangleArea(vector<int>& heights) {
        auto ret{0};
        heights.push_back(0);
        stack<pair<int, int>> st{};
        for (auto i{0}; i < heights.size(); ++i) {
            auto index{i};
            while (!st.empty() && heights[i] < st.top().second) {
                ret = std::max(ret, (i - st.top().first) * st.top().second);
                index = st.top().first;
                st.pop();
            }
            st.push({index, heights[i]});
        }
        return ret;
    }
};
